### PR TITLE
Demonstrate artifact-manager-s3 config

### PR DESCRIPTION
**Pull request by batmat**
[JENKINS-52304](https://issues.jenkins-ci.org/browse/JENKINS-52304)

### CODE REVIEW

File: integrations/src/test/resources/org/jenkinsci/plugins/casc/ArtifactManagerS3.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
+unclassified:
```
Comment (oleg-nenashev) [inline integrations/src/test/resources/org/jenkinsci/plugins/casc/ArtifactManagerS3.yml]: IIUC it is going to change with https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/67

**Comment by oleg-nenashev** [comment]
> because the configuration layout will change once the AWS Category has a Symbol

Maybe it is a compatibility layer we should add @ndeloof @ewelinawilkosz . It looks reasonable to support any global configurations in the "unclassified" category

**Comment by batmat** [comment]
Note that what I am using here is a copy paste of what is *already* merged for Essentials, as per https://github.com/jenkins-infra/evergreen/pull/128 so possibly https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/67 should make it backward compatible.

**Comment by batmat** [comment]
Anyway, the real point of the current PR is the second commit: showing that I have currently no idea on how to make the S3 blob store configured as the global _Artifact Manager_ through CasC.

**Comment by casz** [comment]
From stacktrace try attaching a debugger at `org.jenkinsci.plugins.casc.model.CNode.asMapping(CNode.java:17)`

since it spits out this
```
Caused by: org.jenkinsci.plugins.casc.ConfiguratorException: Item isn't a Mapping
at org.jenkinsci.plugins.casc.model.CNode.asMapping(CNode.java:17)
```

**Comment by batmat** [comment]
@casz thanks! Though I actually do not plan/intend to fix this globally. I miss time to dig into it and have other things to work on in my plate.

To clarify my intent, indeed: I filed this PR hoping hope `artifact-manager-s3` and `configuration-as-code` plugins developers can possibly use this test to find/fix what is missing to make `artifact-manager-s3` fully configurable by CasC.

**Comment by casz** [comment]
@batmat no worries 😅 I'll give it a shot on the test, thanks for the test case!

**Comment by casz** [comment]
JCloudsArtifactManagerFactory expects a BlobStoreProvider ie. S3BlobStore this seems to require a separate configurator unless there is some hidden magic @ndeloof or if https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/67 fixes everything

**Comment by jglick** [comment]
`artifact-manager-s3` is already fully configurable by JEP-201 as shown here, but there are planned changes:

* some sort of refactoring of where general AWS settings live
* addition of symbols to make things look prettier

I expect to take this PR over to reflect those changes. The first is likely incompatible, oh well.

**Comment by ndeloof** [comment]
JCloudsArtifactManagerFactory require a (single) BlobStoreProvider `provider`, not a list

```yaml
  artifactManager:
    artifactManagerFactories:
      - JCloudsArtifactManagerFactory:
          provider : s3blobstoreconfig
```

**Comment by casz** [comment]
I tried `provider: s3blobstoreconfig` but it has to be a `provider` which s3blobstoreconfig is not. The thing that tripped be earlier was I thought the intending was correct.

Passing it a new class of S3BlobStore that gets information from s3blobstoreconfig seems off. What is the correct yaml?

The yaml you provided above gives this stacktrace:
```
java.lang.AssertionError: Unexpected exception 

	at org.jenkinsci.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:38)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:543)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: No io.jenkins.plugins.artifact_manager_jclouds.BlobStoreProvider implementation found for s3blobstoreconfig
	at org.jenkinsci.plugins.casc.HeteroDescribableConfigurator.findDescribableBySymbol(HeteroDescribableConfigurator.java:121)
	at org.jenkinsci.plugins.casc.HeteroDescribableConfigurator.configure(HeteroDescribableConfigurator.java:82)
	at org.jenkinsci.plugins.casc.HeteroDescribableConfigurator.configure(HeteroDescribableConfigurator.java:35)
	at org.jenkinsci.plugins.casc.DataBoundConfigurator.tryConstructor(DataBoundConfigurator.java:150)
	at org.jenkinsci.plugins.casc.DataBoundConfigurator.instance(DataBoundConfigurator.java:63)
	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:199)
	at org.jenkinsci.plugins.casc.DataBoundConfigurator.configure(DataBoundConfigurator.java:98)
	at org.jenkinsci.plugins.casc.HeteroDescribableConfigurator.configure(HeteroDescribableConfigurator.java:85)
	at org.jenkinsci.plugins.casc.HeteroDescribableConfigurator.configure(HeteroDescribableConfigurator.java:35)
	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:265)
	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:200)
	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:270)
	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:200)
	at org.jenkinsci.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:455)
	at org.jenkinsci.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:475)
	at org.jenkinsci.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:380)
	at org.jenkinsci.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:360)
	at org.jenkinsci.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:343)
	at org.jenkinsci.plugins.casc.misc.JenkinsConfiguredWithCodeRule.before(JenkinsConfiguredWithCodeRule.java:35)
```

**Comment by batmat** [comment]
Used in https://github.com/jenkins-infra/evergreen/pull/139

**Comment by ndeloof** [comment]
as https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/67 isn't mergeable for now, shall we just merge as-is @oleg-nenashev and maybe later update the yaml config ?

**Comment by batmat** [comment]
> Please do not merge; I expect to move this into artifact-manager-s3-plugin instead.

I am not sure this is desirable. IMO it's nice, if not required, that casc has a bunch of integration with critical/central plugins tested as things are being developed. It could make sense to do both, but having the `artifact-manager-s3` *integration* tested here makes sense too IMO.

**Comment by batmat** [comment]
@ndeloof resolved the issue, and bumped to latest. 
Can you remove the `work-in-progress` label?

For the fact it should, or should not, be merged, it's another matter I'd like to hear confirmation from people here. Esp. @oleg-nenashev and @jglick who wrote these tests should not be here. I explained in https://github.com/jenkinsci/configuration-as-code-plugin/pull/347#issuecomment-405203870 why I only agree partially.

Thanks everyone.

**Comment by jglick** [comment]
> it's nice, if not required, that casc has a bunch of integration with critical/central plugins tested as things are being developed

I would say this repo should host a small but representative assortment of integration tests with longstanding, widely-used, generic plugins with very stable configurations that would demonstrate particular interesting aspects of the syntax. It is certainly unrealistic for it to host integration tests for every plugin you might want to configure. As a rule a more domain-specific plugin like `artifact-manager-s3` ought to have test deps on a more generic plugin like `configuration-as-code` and not the other way around. Merging this PR does not add new test coverage (we already had that in https://github.com/jenkinsci/artifact-manager-s3-plugin/pull/72) yet adds stuff to maintain in PCT failures if we wind up tweaking the semantics of AWS and/or S3 configuration.

**Comment by ndeloof** [comment]
@jglick I indeed would like to see such integration test hosted by other plugins, anyway this is way too early, we need at least to reach 1.0, and in the meantime such integration tests help to catch regressions. In middle terms future we could indeed only host a subset of integration tests to demonstrate some significant configuration use-cases, either to demonstrate interesting syntax, or to test/showcase some specific implementation / introspection details

**Pull request merged**